Adding `--iree-opt-import-parameters=` and cleaning up export.

benvanik opened the pull request:
The new option allows for importing all or some parameters from a parameter file into the compiler. One or more parameter files to one or more scopes can be specified, just as with the `--parameters=[scope=]file.*` flag in the runtime tool. If a scope is provided only parameters with that scope will be imported. If a parameter (optionally within a scope) is named explicitly it will be imported. If a maximum size is specified all parameters <= that size will be imported.

The export pass has been reworked to be the inverse of the import such that it's possible to round-trip through export and import. The import pass doesn't currently support all of the types we can export as `SerializableAttrInterface` doesn't have a deserialize method (yet) and we need special handling of packed types (i1, i4, etc). For now warnings are emitted and the globals are left as parameter references to still allow partial imports.

Flag renames:
* `--iree-opt-parameter-archive-export-file=` + `--iree-opt-parameter-archive-export-scope=` -> `--iree-opt-export-parameters=[scope=]file`
* `--iree-opt-minimum-parameter-export-size=` -> `--iree-opt-export-parameter-minimum-size=`
* `--iree-opt-splat-parameter-archive-export-file=` -> `--iree-opt-splat-parameters=`

benvanik committed:
Adding `--iree-opt-import-parameters=` option.
This allows for importing all or some parameters from a parameter file
into the compiler. Currently only one import can be specified with
the flag but we could extend that to multiple in the future by following
the same flag conventions as the runtime tooling.

If a scope is provided only parameters with that scope will be imported.
If a parameter (optionally within a scope) is named explicitly it will
be imported. If a maximum size is specified all parameters <= that size
will be imported.

This also renames the export flags, as they are inconsistent.

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/test/import_parameters.mlir:
```diff
@@ -0,0 +1 @@
+// DO NOT SUBMIT
```
Tagging this in a comment to resolve (add tests)

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp:
```diff
@@ -0,0 +1,419 @@
+// Copyright 2023 The IREE Authors
```
nit: if file is new, could use 2024

```suggestion
// Copyright 2024 The IREE Authors
```

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp:
```diff
@@ -0,0 +1,419 @@
+        iree_io_parse_safetensors_index(fileHandle->get(), parameterIndex),
+        "parsing safetensors file");
+  } else {
+    llvm::errs() << "unsupported archive file format: " << path << "\n";
```
nit: wrap strings in quotes when logging them, so empty strings stand out
```suggestion
    llvm::errs() << "unsupported archive file format: '" << path << "'\n";
```

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp:
```diff
@@ -0,0 +1,419 @@
+
+// Today only shaped types of elements where we know we can directly access the
+// data as stored in teh file.
+static bool isTypeSupported(Type type) {
```
nit: typo
```suggestion
// Today only shaped types of elements where we know we can directly access the
// data as stored in the file.
static bool isTypeSupported(Type type) {
```

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp:
```diff
@@ -0,0 +1,419 @@
+      assert(false && "integer width not supported");
+      return failure();
+    }
+    return IntegerAttr::get(globalType, APInt(integerType.getWidth(), value));
```
build errors around here
https://github.com/openxla/iree/actions/runs/8345780110/job/22841503854?pr=16828#step:4:7622
```
 error: no viable conversion from returned value of type 'mlir::IntegerAttr' to function return type 'FailureOr<mlir::TypedAttr>'
    return IntegerAttr::get(globalType, APInt(integerType.getWidth(), value));
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

benvanik committed:
Centralizing parameter archive parsing format selection.

benvanik committed:
Reworking ExportParameterPass to be the inverse of ImportParametersPass.
This also moves the pass to run at the very end of the global opt
pipeline and not in the fixed-point iterator so we only export the
parameters once and only after they've been optimized. The behavior
of the pass was changed to only export global constants and not inline
ones as constant expression hoisting (HoistIntoGlobalsPass) should
be the arbiter of what a parameter is.

ScottTodd commented on runtime/src/iree/io/formats/parser_registry.h:
```diff
@@ -0,0 +1,27 @@
+extern "C" {
+#endif  // __cplusplus
+
+// DO NOT SUBMIT path may be full path or just extension
```
👀

benvanik replied:
whoops

ScottTodd reviewed (commented):
reviewing

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/Passes.td:
```diff
@@ -11,32 +11,49 @@ include "mlir/Pass/PassBase.td"
+  ];
+}
+
+def ImportParametersPass :
+    Pass<"iree-io-import-parameters", "mlir::ModuleOp"> {
+  let summary = "Imports parameters from an archive file.";
```
We should also update https://iree.dev/guides/parameters/#working-with-parameter-files

Guess I can add that to my docs backlog along with updating the profiling / Tracy docs... and soon maybe also the PyTorch docs...

ScottTodd replied:
I could fold doc updates into https://github.com/openxla/iree/pull/17006 if this PR lands first

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/test/import_parameters.mlir:
```diff
@@ -0,0 +1,36 @@
+// RUN: iree-opt --pass-pipeline="builtin.module(iree-io-export-parameters{path="opt=%t.irpa" minimum-size=0},iree-io-import-parameters{paths="opt=%t.irpa"})" %s | FileCheck %s
```
Checking my understanding: this exports _then_ imports, round tripping?

I suppose we could also test with `iree-create-parameters`, if we didn't want to rely on the sequencing there. Feels sneaky to me to have the files needed by the compiler not all already available on disk when the command begins.

benvanik replied:
round-tripping is best - it's more sneaky to have runtime deps in compiler lit tests :)

ScottTodd replied:
Ah, good point. SGTM

ScottTodd reviewed (commented):
Code changes LGTM. A few CI failures still + one `DO NOT SUBMIT` in a comment

benvanik committed:
Fixing global opt pipeline info.
